pki/leaf.go

```diff
@@ -0,0 +1,244 @@
+// Copyright 2020 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package pki
+
+import (
+	"bytes"
+	"crypto"
+	"crypto/tls"
+	"crypto/x509"
+	"crypto/x509/pkix"
+	"net"
+	"sort"
+
+	"github.com/juju/errors"
+)
+
+type DefaultLeaf struct {
+	group          string
+	certificate    *x509.Certificate
+	chain          []*x509.Certificate
+	signer         crypto.Signer
+	tlsCertificate *tls.Certificate
+}
+
+type DefaultLeafRequest struct {
+	dnsNames      map[string]struct{}
+	ipAddresses   map[string]net.IP
+	leafMaker     LeafMaker
+	requestSigner CertificateRequestSigner
+	signer        crypto.Signer
+	subject       pkix.Name
+}
+
+type Leaf interface {
+	// Certificate returns the x509 certificate of this leaf. May be nil if no
+	// certificate exists yet. Call Commit to sign the leaf.
+	Certificate() *x509.Certificate
+
+	// Chain is the certificate signing chain for this leaf in the case of
+	// intermediate CA's
+	Chain() []*x509.Certificate
+
+	Signer() crypto.Signer
+
+	TLSCertificate() *tls.Certificate
+
+	ToPemParts() (cert, key []byte, err error)
+}
+
+type LeafMaker func(*x509.Certificate, []*x509.Certificate, crypto.Signer) (Leaf, error)
+
+type LeafRequest interface {
+	AddDNSNames(...string) LeafRequest
+	AddIPAddresses(...net.IP) LeafRequest
+	Commit() (Leaf, error)
+}
+
+var (
+	HeaderLeafGroup = "leaf.pki.juju.io/group"
```

Comment:
juju.is not juju.io
juju.io is some k8s legacy only

Reply:
Last PR had me back this out. Need some consistency